File: src/core/namespace.ts
import { mkdir } from "node:fs/promises";
import { join } from "node:path";

/**
 * Chat history and memory get a directory boundary per workspace, not a
 * shared table/file with a workspace column -- easier to verify by
 * inspection than a query filter.
 */
export function chatHistoryDir(dataDir: string, workspaceId: string): string {
  return join(dataDir, "workspaces", workspaceId, "chat");
}

export function memoryDir(dataDir: string, workspaceId: string): string {
  return join(dataDir, "workspaces", workspaceId, "memory");
}

export async function ensureWorkspaceDirs(dataDir: string, workspaceId: string): Promise<void> {
  await mkdir(chatHistoryDir(dataDir, workspaceId), { recursive: true });
  await mkdir(memoryDir(dataDir, workspaceId), { recursive: true });
}

File: python/src/workspaceguard/namespace.py
"""
Chat history and memory get a directory boundary per workspace, not a
shared table/file with a workspace column -- easier to verify by inspection
than a query filter. Ported from src/core/namespace.ts.
"""
from __future__ import annotations

import os


def chat_history_dir(data_dir: str, workspace_id: str) -> str:
    return os.path.join(data_dir, "workspaces", workspace_id, "chat")


def memory_dir(data_dir: str, workspace_id: str) -> str:
    return os.path.join(data_dir, "workspaces", workspace_id, "memory")


async def ensure_workspace_dirs(data_dir: str, workspace_id: str) -> None:
    os.makedirs(chat_history_dir(data_dir, workspace_id), exist_ok=True)
    os.makedirs(memory_dir(data_dir, workspace_id), exist_ok=True)
